pkg/apis/batch/validation/validation.go

```diff
@@ -179,10 +202,103 @@ func validateJobSpec(spec *batch.JobSpec, fldPath *field.Path, opts apivalidatio
 			spec.Template.Spec.RestartPolicy, []string{string(api.RestartPolicyOnFailure), string(api.RestartPolicyNever)}))
+	} else if spec.PodFailurePolicy != nil && spec.Template.Spec.RestartPolicy != api.RestartPolicyNever {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("template", "spec", "restartPolicy"),
+			spec.Template.Spec.RestartPolicy, fmt.Sprintf("only %q is supported when podFailurePolicy is specified", api.RestartPolicyNever)))
 	}
 	return allErrs
 }
 
+func validatePodFailurePolicy(spec *batch.JobSpec, fldPath *field.Path, allErrs field.ErrorList) field.ErrorList {
+	containerNames := sets.NewString()
+	for _, containerSpec := range spec.Template.Spec.Containers {
+		containerNames.Insert(containerSpec.Name)
+	}
+	for _, containerSpec := range spec.Template.Spec.InitContainers {
+		containerNames.Insert(containerSpec.Name)
+	}
+	for i, rule := range spec.PodFailurePolicy.Rules {
+		rulePath := fldPath.Child("podFailurePolicy").Index(i)
+
+		actionPath := rulePath.Child("action")
+		if rule.Action == "" {
+			allErrs = append(allErrs, field.Required(actionPath, fmt.Sprintf("valid values: %q", supportedPodFailurePolicyActions.List())))
+		} else if !supportedPodFailurePolicyActions.Has(string(rule.Action)) {
+			allErrs = append(allErrs, field.NotSupported(actionPath, rule.Action, supportedPodFailurePolicyActions.List()))
+		}
+		if rule.OnExitCodes != nil {
+			onExitCodesPath := rulePath.Child("onExitCodes")
+			operatorPath := onExitCodesPath.Child("operator")
+			if rule.OnExitCodes.Operator == "" {
+				allErrs = append(allErrs, field.Required(operatorPath, fmt.Sprintf("valid values: %q", supportedPodFailurePolicyOnExitCodesOperator.List())))
+			} else if !supportedPodFailurePolicyOnExitCodesOperator.Has(string(rule.OnExitCodes.Operator)) {
+				allErrs = append(allErrs, field.NotSupported(operatorPath, rule.OnExitCodes.Operator, supportedPodFailurePolicyOnExitCodesOperator.List()))
+			}
+			if rule.OnExitCodes.ContainerName != nil && !containerNames.Has(*rule.OnExitCodes.ContainerName) {
+				path := onExitCodesPath.Child("containerName")
+				allErrs = append(allErrs, field.Invalid(path, *rule.OnExitCodes.ContainerName, "must be one of the container names in the pod template"))
+			}
+			valuesPath := onExitCodesPath.Child("values")
+			if len(rule.OnExitCodes.Values) == 0 {
+				allErrs = append(allErrs, field.Invalid(valuesPath, rule.OnExitCodes.Values, "at least one value is required"))
+			} else {
+				isOrdered := true
+				uniqueValues := sets.NewInt32()
+				for j, exitCodeValue := range rule.OnExitCodes.Values {
+					if rule.OnExitCodes.Operator == batch.PodFailurePolicyOnExitCodesOpIn && exitCodeValue == 0 {
+						valuePath := valuesPath.Index(j)
+						allErrs = append(allErrs, field.Invalid(valuePath, exitCodeValue, "must not be 0 for the In operator"))
+					}
+					uniqueValues.Insert(rule.OnExitCodes.Values[j])
```

Comment:
Before inserting, you could check if the value is already in the set. Then you can pin point exactly which element of the array is repeated in the error message.

Reply:
Updated, pin point by values rather than indexes. It seems having values should be enough. We could also report indexes, but then we should probably update also the index of the first element (which was not yet considered as duplicate).

Reply:
Updated again, there is an already exiting error method to report duplicates. Using it.